pkg/cli/mirror/catalog_images.go

```diff
@@ -233,6 +280,167 @@ func (o *MirrorOptions) processCatalogRefs(ctx context.Context, catalogsByImage
+		layoutPath := layout.Path(v1alpha2.TrimProtocol(srcRef.OCIFBCPath))
+
+		// get its index.json and obtain its manifest
+		rootIndex, err := layoutPath.ImageIndex()
+		if err != nil {
+			return err
+		}
+		rootIndexManifest, err := rootIndex.IndexManifest()
+		if err != nil {
+			return err
+		}
+
+		// attempt to find the first image reference in the layout...
+		// for a manifest list only search one level deep.
+
+	loop:
+		for _, descriptor := range rootIndexManifest.Manifests {
+
+			if descriptor.MediaType.IsIndex() {
+				// follow the descriptor using its digest to get the referenced index and its manifest
+				childIndex, err := rootIndex.ImageIndex(descriptor.Digest)
+				if err != nil {
+					return err
+				}
+				childIndexManifest, err := childIndex.IndexManifest()
+				if err != nil {
+					return err
+				}
+
+				// at this point, find the first image and store it for later if possible
+				//TODO extract the child index that corresponds to this machine's architecture
+				for _, childDescriptor := range childIndexManifest.Manifests {
+					if childDescriptor.MediaType.IsImage() && childDescriptor.Platform.Architecture == runtime.GOARCH && childDescriptor.Platform.OS == runtime.GOOS {
+						img, err = childIndex.Image(childDescriptor.Digest)
+						if err != nil {
+							return err
+						}
+						// no further processing necessary
+						break loop
+					}
+				}
+
+			} else if descriptor.MediaType.IsImage() {
+				// this is a direct reference to an image, so just store it for later
+				img, err = rootIndex.Image(descriptor.Digest)
+				if err != nil {
+					return err
+				}
+				// no further processing necessary
+				break loop
+			}
+		}
+	}
+	// if we get here and no image was found bail out
+	if img == nil {
+		return fmt.Errorf("unable to obtain image for %v", srcRef)
+	}
+	tr := tar.NewReader(mutate.Extract(img))
+	opmBinaryExtracted := ""
+	for {
+		header, err := tr.Next()
+
+		// break the infinite loop when EOF
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		// skip the file if it is a directory or not in the bin dir
+		if !strings.HasSuffix(header.Name, opmBinarySuffix) || header.FileInfo().IsDir() {
+			continue
+		}
+
+		var buf bytes.Buffer
+		_, err = buf.ReadFrom(tr)
+		if err != nil {
+			return err
+		}
+
+		targetFileName := filepath.Join(outDir, header.Name)
+		bytes := buf.Bytes()
+
+		baseDir := filepath.Dir(targetFileName)
+		err = os.MkdirAll(baseDir, 0755)
+		if err != nil {
+			return err
+		}
+
+		f, err := os.Create(targetFileName)
+		if err == nil {
+			defer f.Close()
+		} else {
+			return err
+		}
+
+		_, err = f.Write(bytes)
+		if err != nil {
+			return err
+		}
+		opmBinaryExtracted = targetFileName
+	}
+	// check for the folder (it should exist if we found something)
+	_, err = os.Stat(opmBinaryExtracted)
+	if errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("opm binary not found after extracting %q from catalog image %v", opmBinaryExtracted, srcRef)
+	}
```

Comment:
Since you're looping through the tar file and finding potentially more than one opm file, this is effectively only checking the last one you found. Maybe that's OK? Also the comment says a folder, but it actually checking a file